### PR TITLE
#180822215 Create an Input Component

### PR DESCRIPTION
**Pull request by RACHID-WEBDEV**
#### What does this PR do?
`Create a reuseable Input Component`
#### Description of Task to be completed?
- An Input component that can be passed to several other components
- Added and `SVG Icon` props to make the input reusable with or without an Icon
#### How should this be manually tested?
- Clone the Repository
- Run `yarn Install` in your terminal to install required dependencies
-  Run `yarn dev` to start the local development server 
- Open your browser and check `localhost:3000`

### CODE REVIEW

**Review by andela-hpopoola — CHANGES_REQUESTED** [review]
Thank you for working on this. It looks like you have combined the tasks here with the Storybook stories. Can you try removing the unrelated files so that this contains only changes for Input Component?

As I explained to you on the call today, when you are working on a new unrelated story, it is best to follow this workflow: 
- checkout in the `main` branch
- use `git pull` to pull the updated content
- use `git checkout -b <BRANCH_NAME>` to create and checkout into a new branch, lets call this branch `button-branch`
- make your changes, commit, push to GitHub and create your PR


After you have finished work on the `button-branch`, if you need to do a new story on `Input` which is not related to `button-branch`, you will need to checkout into the main branch again and repeat the process above as explained below:
- I am assuming you are still in the `button-branch`
- You will need to checkout back to the `main` branch, as this is the base branch and it contains only approved codes.
- In the `main` branch, use `git pull` to pull the updated content (this is necessary incase a new PR has been merged into the code)
- use `git checkout -b input-story` to create and checkout into a new branch for input called `input-story`.
- make your changes, commit, push to GitHub and create your PR

After you have finished working on the `input-story` branch, and you need to make changes on the `button-branch`, then you will need to checkout into this branch, make your changes and push back to GitHub (the same applies to the `input-story` branch).

I hope this clears up the mix up. Please let me know if you have any questions on this.

**Comment by RACHID-WEBDEV** [comment]
#### Yes  getting more clearer and getting more intresting

File: assets/svgs.js [hunk 1]
```diff
@@ -0,0 +1,7 @@
+export const EmailIcon = (props) => {
```
Comment (andela-hpopoola) [inline assets/svgs.js:1]: The email Icon is not part of the design guideline. Can we replace it with the Password Icon 🙏🏽 ?

File: components/form/Input.js [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+          name={name}
+          placeholder={placeholder}
+          type={type}
+        />
+        <div className="absolute inset-y-0 right-0 flex items-center px-3 pointer-events-none">{Icon}</div>
+      </div>
+    </div>
+  );
+};
+Input.propTypes = {
```
Comment (andela-hpopoola) [inline components/form/Input.js:20]: Can you try adding some reasonable defaults where necessary to minimize forcing the user to use all the fields?

File: components/form/Input.js [hunk 3]
```diff
@@ -0,0 +1,26 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
+const Input = ({ Icon, label, name, placeholder, type }) => {
```
Comment (andela-hpopoola) [inline components/form/Input.js:4]: Can we add className to the props too to support additional customization?

File: components/form/Input.js [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    <div className="w-1/3">
+      <div className="relative text-[#344055]">
+        <input
+          className="w-full h-10 pl-3 pr-8 text-base bg-[#D6D9DD] placeholder-[#344055] border rounded-md focus:shadow-outline"
+          label={label}
+          name={name}
+          placeholder={placeholder}
+          type={type}
+        />
+        <div className="absolute inset-y-0 right-0 flex items-center px-3 pointer-events-none">{Icon}</div>
```
Comment (andela-hpopoola) [inline components/form/Input.js:15]: The Icon should be an optional prop (see the email input field in the mockup) so you might want to cater for this too.

**Review by andela-hpopoola — CHANGES_REQUESTED** [review]
Thank you for working on this. I left a few comments on ensuring the Input field is properly implemented. This should be good to go when you fix the comments.

File: assets/svgs.js (hunk 1 above)
Comment (RACHID-WEBDEV) [reply assets/svgs.js:1]: okay sir noted

**Review by andela-hpopoola — CHANGES_REQUESTED** [review]
Thank you for working on this. Can you try to fix the remaining comments and fix the merge conflict so that we can resolve this story.

**Review by andela-hpopoola — CHANGES_REQUESTED** [review]
Thank you for working on this. I dropped a few comments and also can you try to fix the existing comments I left on this previously.

Lastly, I have fixed the merge conflict, so it should be easier to you to continue work. You might want to delete the existing `ft-Create-an-Input-Component-#180822215` branch on your local machine and pull the current one from GitHub to save you time. 

- https://www.freecodecamp.org/news/how-to-delete-a-git-branch-both-locally-and-remotely/
- https://stackoverflow.com/a/1783426

File: components/form/Input.js [hunk 5]
```diff
@@ -21,7 +21,12 @@ const Input = ({ Icon, label, name, type }) => {
 };
 Input.propTypes = {
   label: PropTypes.string,
-  name: PropTypes.string,
+  name: PropTypes.string.required,
```
Comment (andela-hpopoola) [inline components/form/Input.js:24]: I think this should be `isRequired`:

https://reactjs.org/docs/typechecking-with-proptypes.html

```suggestion
  name: PropTypes.string.isRequired,
```

**Review by andela-hpopoola — CHANGES_REQUESTED** [review]

File: components/form/Input.js [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+        <label className="text-sm absolute top-11 text-[#344055] left-2 pointer-events-none transition-all">
+          {label}
+        </label>
+        <span className="absolute top-12 right-2 cursor-pointer ">{Icon}</span>
+      </div>
+    </div>
+  );
+};
+Input.propTypes = {
+  label: PropTypes.string,
```
Comment (andela-hpopoola) [inline components/form/Input.js:23]: We need to add the Icon props here too.

**Review by andela-hpopoola — CHANGES_REQUESTED** [review]
Thank you for the hard work on this. I have resolved (or closed) the comments that you have fixed. You will need to fix the pending comments to enable us to merge this. Here are the ones I noticed:

1. Add className props to the Input field to support additional customization
2. Use the label text as the placeholder name so that we will not need to use placeholder when it is the same as the label text.
3. Use `name: PropTypes.string.isRequired,` to state that the name props is required - Here is the [documentation](https://reactjs.org/docs/typechecking-with-proptypes.html).
4. We need to add the `Icon` props to the prop types used for Input component and ensure that nothing is rendered when the Icon props is not given.

Thank you for the hard work once again. Feel free to reach out if you don't understand any of the things stated above or you have further questions. Nice work.

**Comment by andela-hpopoola** [comment]
Thank you for the updates, however, it looks like you didn't fix all of the issues I raised 😞. Can you refer back to my comment here: https://github.com/Amcovad/amcovad-frontend/pull/8#pullrequestreview-853890316


>   1. Add className props to the Input field to support additional customization
>  2. Use the label text as the placeholder name so that we will not need to use placeholder when it is the same as the label text.
>  3. Use `name: PropTypes.string.isRequired,` to state that the name props is required - Here is the [documentation](https://reactjs.org/docs/typechecking-with-proptypes.html).
>  4. We need to add the `Icon` props to the prop types used for Input component and ensure that nothing is rendered when the Icon props is not given.

**Comment by andela-hpopoola** [comment]
Thank you for the hard work on this. This is finally good to go 🎉 🎉 🎉 . Feel free to merge.

**Review by andela-hpopoola — APPROVED** [review]

**Pull request merged**